### PR TITLE
Implement Proxy And Fix m-team

### PR DESCRIPTION
**Pull request by usnail**
#### Database Migration
NO

#### Description
Implement Indexer Web Proxy
Inject web proxy when FlareSolverr choosed

#### Issues Fixed or Closed by this PR

* Fixed: (MTeamTP2FA/m-team) redirect error

### CODE REVIEW

**Comment by bakerboy448** [comment]
So the issue with m team is they just need a useragent override? That should and can be done in the definition rather than all this.

also there's numerous lines being changed/reverted in cloudflare for no apparent reason?

**Comment by usnail** [comment]
> So the issue with m team is they just need a useragent override? That should and can be done in the definition rather than all this.
> 
> also there's numerous lines being changed/reverted in cloudflare for no apparent reason?

yes m team just need a useragent override, and mybe allow some page redirect

I found that indexers could not use  index proxy (like web proxy ) by tag 
[12be8bb](https://github.com/Prowlarr/Prowlarr/pull/1103/commits/12be8bbf6b45fc9ac9f4baf01877e85186b6c3e4)

Some protect   Site by cloudflare mybe need proxy , when choosed FlareSolverr  and web proxy tags  will transfer proxy info to FlareSolverr  
f659bf7bb118d8892608f8c8d65ed975ba6b345f

**Comment by usnail** [comment]
> > So the issue with m team is they just need a useragent override? That should and can be done in the definition rather than all this.
> > also there's numerous lines being changed/reverted in cloudflare for no apparent reason?
> 
> yes m team just need a useragent override, and mybe allow some page redirect
> 
> I found that indexers could not use index proxy (like web proxy ) by tag [12be8bb](https://github.com/Prowlarr/Prowlarr/pull/1103/commits/12be8bbf6b45fc9ac9f4baf01877e85186b6c3e4)
> 
> Some protect Site by cloudflare mybe need proxy , when choosed FlareSolverr and web proxy tags will transfer proxy info to FlareSolverr [f659bf7](https://github.com/Prowlarr/Prowlarr/commit/f659bf7bb118d8892608f8c8d65ed975ba6b345f)

InjectCookies method Covered by old code

**Comment by ta264** [comment]
Proxies in general and flaresolver is particular already work I believe. Need some justification as to why these changes are needed.

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs [hunk 1]
```diff
@@ -112,9 +134,9 @@ private HttpRequest GenerateFlareSolverrRequest(HttpRequest request)
             }
             else if (request.Method == HttpMethod.Post)
             {
-                var contentTypeType = request.Headers.ContentType.ToLower() ?? "<null>";
+                var contentTypeType = request.Headers.ContentType;
```
Comment (bakerboy448) [inline src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:137]: Why is this reverting one of the recent commits that fixed Flaresolver issues?

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs [hunk 2]
```diff
@@ -112,9 +134,9 @@ private HttpRequest GenerateFlareSolverrRequest(HttpRequest request)
...
 
-                if (contentTypeType.Contains("application/x-www-form-urlencoded"))
+                if (contentTypeType == "application/x-www-form-urlencoded")
```
Comment (bakerboy448) [inline src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:139]: ditto

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs [hunk 3]
```diff
@@ -133,8 +155,7 @@ private HttpRequest GenerateFlareSolverrRequest(HttpRequest request)
                         UserAgent = userAgent
                     };
                 }
-                else if (contentTypeType.Contains("multipart/form-data")
-                         || contentTypeType.Contains("text/html"))
+                else if (contentTypeType.Contains("multipart/form-data"))
```
Comment (bakerboy448) [inline src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:158]: Same here

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs [hunk 4]
```diff
@@ -149,15 +170,16 @@ private HttpRequest GenerateFlareSolverrRequest(HttpRequest request)
...
             }
 
+            InjectProxy(req);
+
             var apiUrl = string.Format("{0}/v1", Settings.Host.TrimEnd('/'));
             var newRequest = new HttpRequest(apiUrl, HttpAccept.Json);
 
             newRequest.Headers.ContentType = "application/json";
             newRequest.Method = HttpMethod.Post;
-            newRequest.LogResponseContent = true;
```
Comment (bakerboy448) [inline src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:157]: Same here

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs [hunk 5]
```diff
@@ -149,15 +170,16 @@ private HttpRequest GenerateFlareSolverrRequest(HttpRequest request)
...
             newRequest.SetContent(req.ToJson());
 
-            _logger.Debug("Cloudflare Detected, Applying FlareSolverr Proxy {0} to request {1}", Name, request.Url);
+            _logger.Debug("Applying FlareSolverr Proxy {0} to request {1}", Name, request.Url);
```
Comment (bakerboy448) [inline src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:182]: Same here

File: src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public HttpResponse PostResponse(HttpResponse response)
+        {
+            var request = response.Request;
+
+            // some times,site always redirect to verify page even though has benn already login done
+            if (request.Url.Host == Host && response.HasHttpRedirect && response.RedirectUrl.Contains("verify.php", StringComparison.Ordinal))
+            {
+                request.AllowAutoRedirect = true;
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs:30]: Redirects can already be allowed / followed with a definition flag

File: src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+using System;
+using NzbDrone.Common.Http;
+
+namespace NzbDrone.Core.Http
+{
+    public class MTeamTpHttpRequestInterceptor : IHttpRequestInterceptor
+    {
+        private string Host => "kp.m-team.cc";
+
+        private string UserAgent => "Mozilla/5.0 (X11; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0";
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs:10]: Useragent can already be set via the definition

**Comment by usnail** [comment]
> Proxies in general and flaresolver is particular already work I believe. Need some justification as to why these changes are needed.

I can create different indexer proxy apply indexes  use tags, have a higher priority
[general can](https://github.com/Prowlarr/Prowlarr/pull/1103/commits/12be8bbf6b45fc9ac9f4baf01877e85186b6c3e4)

File: src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs (hunk 7 above)
Comment (usnail) [reply src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs:10]: fix m team only file https://github.com/Prowlarr/Prowlarr/pull/1103/files/f659bf7bb118d8892608f8c8d65ed975ba6b345f#diff-fc204774368bcfdaddf157f615f145fec411d3f65f2e83f61b4e5748cfca20cb

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs (hunk 4 above)
Comment (usnail) [reply src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:157]: The files got corrupted to binary after my computer restarted

So I copied the old changes

File: src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs (hunk 5 above)
Comment (usnail) [reply src/NzbDrone.Core/IndexerProxies/FlareSolverr/FlareSolverr.cs:182]: The files got corrupted to binary after my computer restarted

So I copied the old changes

File: src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs (hunk 6 above)
Comment (usnail) [reply src/NzbDrone.Core/Http/MTeamTpHttpRequestInterceptor.cs:30]: I can't find how to modify the definition flag and didn't know just verify page  required redirect

**Comment by bakerboy448** [comment]
Conflicted

also proxy logic hopefully fixed with https://github.com/Prowlarr/Prowlarr/commit/f97b35403d24e752f5a2442056d5cdd3e155d283

**Pull request closed without merging**